src/World.cpp

```diff
@@ -572,50 +572,147 @@ void cWorld::Start(void)
+		if ((Biome == EMCSBiome::biOcean) || (Biome == EMCSBiome::biFrozenOcean))
+		{
+			BiomeOffset += Vector3d(cChunkDef::Width * 4, 0, 0);
+			continue;
+		}
 
-		if (
-			(biome != biOcean) && (biome != biFrozenOcean) &&  // The biome is acceptable (don't want a small ocean island)
-			!IsBlockWaterOrIce(GetBlock(SpawnX, GetHeight(SpawnX, SpawnZ), SpawnZ))  // The terrain is acceptable (don't want to spawn inside a lake / river)
-		)
+		// Found a usable biome
+		// Spawn chunks so we can find a nice spawn.
+		int ChunkX = 0, ChunkZ = 0;
+		cChunkDef::BlockToChunk(BiomeOffset.x, BiomeOffset.z, ChunkX, ChunkZ);
+		cSpawnPrepare::PrepareChunks(*this, ChunkX, ChunkZ, a_MaxSpawnRadius);
+		break;
+	} while ((--BiomeCheckCount) > 0);
+	
+	// Check 0, 0 first.
+	double SpawnY = 0.0;
+	if (CanSpawnAt(BiomeOffset.x, BiomeOffset.z, SpawnY))
+	{
+		m_SpawnX = BiomeOffset.x + 0.5;
+		m_SpawnY = SpawnY;
+		m_SpawnZ = BiomeOffset.z + 0.5;
+
+		LOGINFO("Generated spawnpoint position at {%i, %i, %i}", static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnY), static_cast<int>(m_SpawnZ));
+		return;
+	}
+
+	// A search grid (searches clockwise around the origin)
+	static const Vector3i ChunkOffset[] =
+	{
+		Vector3i(0, 0, cChunkDef::Width),
+		Vector3i(cChunkDef::Width, 0, cChunkDef::Width),
+		Vector3i(cChunkDef::Width, 0, 0),
+		Vector3i(cChunkDef::Width, 0, -cChunkDef::Width),
+		Vector3i(0, 0, -cChunkDef::Width),
+		Vector3i(-cChunkDef::Width, 0, -cChunkDef::Width),
+		Vector3i(-cChunkDef::Width, 0, 0),
+		Vector3i(-cChunkDef::Width, 0, cChunkDef::Width),
+	};
+
+	static const int PerRadiSearchCount = ARRAYCOUNT(ChunkOffset);
+
+	for (int RadiusOffset = 1; RadiusOffset < a_MaxSpawnRadius; ++RadiusOffset)
+	{
+		for (int SearchGridIndex = 0; SearchGridIndex < PerRadiSearchCount; ++SearchGridIndex)
+		{
+			const Vector3i PotentialSpawn = BiomeOffset + (ChunkOffset[SearchGridIndex] * RadiusOffset);
+			
+			if (CanSpawnAt(PotentialSpawn.x, PotentialSpawn.z, SpawnY))
+			{
+				m_SpawnX = PotentialSpawn.x + 0.5;
+				m_SpawnY = SpawnY;
+				m_SpawnZ = PotentialSpawn.z + 0.5;
+
+				LOGINFO("Generated spawnpoint position at {%i, %i, %i}", static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnY), static_cast<int>(m_SpawnZ));
+				return;
+			}
+		}
+	}
+
+	m_SpawnY = GetHeight(static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnZ));
+	LOGWARNING("Did not find an acceptable spawnpoint. Generated a random spawnpoint position at {%i, %i, %i}", static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnY), static_cast<int>(m_SpawnZ));
+}
+
+
+
+
+
+bool cWorld::CanSpawnAt(double a_X, double a_Z, double &a_Y)
+{
+	static const BLOCKTYPE ValidSpawnBlocks[] =
+	{
+		E_BLOCK_GRASS,
+		E_BLOCK_DIRT,
+		E_BLOCK_SAND,
+		E_BLOCK_SNOW,
+		E_BLOCK_SNOW_BLOCK,
+		E_BLOCK_NETHERRACK
+	};
+
+	static const int ValidSpawnBlocksCount = ARRAYCOUNT(ValidSpawnBlocks);
+
+	for (int PotentialY = 96; PotentialY > 32; --PotentialY)
```

Comment:
If a user would use /spawn he should be send back to the linked overworld, or the default world. That's something the Core should fix.

Keep in mind that the height of the nether can be changed in Cuberite. By default it's 128, but you can set it as high or low as you want.

I believe vanilla picks a "random" (based on the seed) location between something like 500 and -500 for both X and Z coordinates. From there it searches for a valid spot. Or at least that's how I always thought it worked ;)

With this there is a chance that you can spawn in a cave isn't there?